File: 220822-kNumber/app/src/main/java/KNumber.java
import java.util.*;

// 2~5번째 까지 자른다 -> array[2 - 1 ] ~ array[ 5 - 1 ]
// sort 오름차순 정열
// array로 만듬 -> 식으로 리팩토링
public class KNumber {
  public int[] solution(int[] array, int[][] commands) {

    array = new int[] {1, 5, 2, 6, 3, 7, 4};
    int[] answer = new int [commands.length];

    for(int i = 0 ; i < commands.length; i += 1) {

      int[] slicedArray = sliceArray(array,commands[i]);

      answer[i] = slicedArray[commands[i][2] - 1];
      System.out.println("정답: " + answer[i]);
    }

    return answer;
  }

  public int[] sliceArray(int[] array,int[] command) {
    int i = command[0];
    int j = command[1];
    int k = command[2];

    int[] slicedArray = new int[j - (i - 1) ];
    int z = 0;
    for(int m = i -1 ; m < j; m += 1) {
      slicedArray[z] = array[m];
      z+=1;
    }
    Arrays.sort(slicedArray);
    return slicedArray;
  }
}

File: 220823-kNumber/kNumber.js
/* eslint-disable import/prefer-default-export */
// 1. 자르기
// 2. 배열 오름차순 정렬
// 3. 나온 배열에서 k번재 숫자 가져오기
// -> 일반화시키기
// 2차원 배열에 적용
export function solution(array, commands) {
  const answer = [];
  let r = 0;
  for (let e = 0; e < commands.length; e += 1) {
    answer[r] = cutSortAndPick(array, commands[e]);
    r += 1;
  }

  return answer;
}

export function cutArray(array, i, j) {
  const slicedArray = [];
  let w = 0;
  for (let q = i - 1; q < j; q += 1) {
    slicedArray[w] = array[q];
    w += 1;
  }

  return slicedArray;
}

export function sortArray(slicedArray) {
  return slicedArray.sort((a, b) => a - b);
}

export function pick(sortedArray, k) {
  return sortedArray[k - 1];
}

export function cutSortAndPick(array, command) {
  const slicedArray = cutArray(array, command[0], command[1]);
  const sortedArray = sortArray(slicedArray);
  const wantedNumber = pick(sortedArray, command[2]);
  return wantedNumber;
}

export function solution2(array, commands) {
  return commands.map((v) => array.slice(v[0] - 1, v[1]).sort((a, b) => a - b).slice(v[2] - 1, v[2])[0]);
}
